contracts/container/contract.go

```diff
@@ -491,6 +494,143 @@ func List(owner []byte) [][]byte {
 	return list
 }
 
+// AddNextEpochNodes accumulates passed nodes as container members for the next
+// epoch to be committed using [CommitContainerListUpdate]. Nodes must be
+// grouped by selector index from placement policy (SELECT clauses). Results of
+// the call operation can be received via [Nodes]. This method must be called
+// only when a container list is changed, otherwise nothing should be done.
+// Call must be signed by the Alphabet nodes.
+func AddNextEpochNodes(cID interop.Hash256, placementVector int, publicKeys []interop.PublicKey) {
+	if len(cID) != interop.Hash256Len {
+		panic(cst.ErrorInvalidContainerID + ": length: " + std.Itoa10(len(cID)))
+	}
+
+	ctx := storage.GetContext()
+	multiaddr := common.AlphabetAddress()
+	common.CheckAlphabetWitness(multiaddr)
+
+	commonPrefix := append([]byte{nextEpochNodesPrefix}, cID...)
+	commonPrefix = append(commonPrefix, byte(placementVector))
+
+	counter := 0
+	c := storage.Find(ctx, commonPrefix, storage.KeysOnly|storage.Backwards)
+	if iterator.Next(c) {
+		counterRaw := iterator.Value(c).([]byte)[1+interop.Hash256Len:]
+		counter = counterFromBE(counterRaw)
+	}
+
+	for _, publicKey := range publicKeys {
+		if len(publicKey) != interop.PublicKeyCompressedLen {
+			panic(cst.ErrorInvalidPublicKey + ": length: " + std.Itoa10(len(publicKey)))
+		}
+
+		counter++
+
+		storageKey := append(commonPrefix, counterBE(counter)...)
+		storage.Put(ctx, storageKey, publicKey)
+	}
+}
+
+func counterBE(c int) []byte {
+	rawCounter := std.Serialize(c)
+	res := []byte{rawCounter[0], rawCounter[1]} // first is type, second is length
+	for i := len(rawCounter) - 1; i > 1; i-- {  // LE to BE
+		res = append(res, rawCounter[i])
+	}
+
+	return res
+}
+
+func counterFromBE(b []byte) int {
+	res := []byte{b[0], b[1]}         // first is type, second is length
+	for i := len(b) - 1; i > 1; i-- { // BE to LE
+		res = append(res, b[i])
+	}
+
+	return std.Deserialize(res).(int)
+}
+
+// CommitContainerListUpdate commits container list changes made by
+// [AddNextEpochNodes] calls in advance. Replicas must correspond to
+// ordered placement policy (REP clauses). If no [AddNextEpochNodes]
+// have been made, it clears container list. Makes "ContainerUpdate"
+// notification with container ID after successful list change.
+// Call must be signed by the Alphabet nodes.
+func CommitContainerListUpdate(cID interop.Hash256, replicas []uint8) {
```

Comment:
Better do `[]int`, we don't really separate int types now.

Reply:
why not considrer it as a part of the interface then? it explains what we expect (even it it does not work)

Reply:
It's the same Integer type in the manifest anyway. Although for NEP-25 there can be some difference, so let's keep it.